core/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java

```diff
@@ -111,49 +162,228 @@ public FieldSortBuilder sortMode(String sortMode) {
+            result.order(SortOrder.readOrderFrom(in));
+        }
+        result.sortMode(in.readOptionalString());
+        result.unmappedType(in.readOptionalString());
+        return result;
+    }
+
+    @Override
+    public FieldSortBuilder fromXContent(QueryParseContext context, String elementName) throws IOException {
+        XContentParser parser = context.parser();
+
+        String fieldName = null;
+        QueryBuilder nestedFilter = null;
+        String nestedPath = null;
+        Object missing = null;
+        SortOrder order = null;
+        String sortMode = null;
+        String unmappedType = null;
+
+        String currentFieldName = null;
+        XContentParser.Token token;
+        fieldName = elementName;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+                while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                    if (token == XContentParser.Token.FIELD_NAME) {
+                        currentFieldName = parser.currentName();
+                    } else if (token == XContentParser.Token.START_OBJECT) {
+                        if (context.parseFieldMatcher().match(currentFieldName, NESTED_FILTER)) {
+                            nestedFilter = context.parseInnerQueryBuilder();
+                        }
+                    } else if (token.isValue()) {
+                        if (context.parseFieldMatcher().match(currentFieldName, NESTED_PATH)) {
+                            nestedPath = parser.text();
+                        } else if (context.parseFieldMatcher().match(currentFieldName, MISSING)) {
+                            missing = parser.objectBytes();
+                        } else if (context.parseFieldMatcher().match(currentFieldName, ORDER)) {
+                            String sortOrder = parser.text();
+                            if ("asc".equals(sortOrder)) {
+                                order = SortOrder.ASC;
+                            } else if ("desc".equals(sortOrder)) {
+                                order = SortOrder.DESC;
+                            } else {
+                                throw new IllegalStateException("Sort order " + sortOrder + " not supported.");
+                            }
+                        } else if (context.parseFieldMatcher().match(currentFieldName, SORT_MODE)) {
+                            sortMode = parser.text();
+                        } else if (context.parseFieldMatcher().match(currentFieldName, UNMAPPED_TYPE)) {
+                            unmappedType = parser.text();
+                        }
+                    }
+                }
+            }
+        }
+
+        FieldSortBuilder builder = new FieldSortBuilder(fieldName);
+        if (nestedFilter != null) {
+            builder.setNestedFilter(nestedFilter);
```

Comment:
Wondering if it would be possible to create the builder first, then call all these setters in the parsing loop above already. Not really that important though.